Coarsen and overwritebcs modifications

joanibal opened the pull request:
## Purpose
- Modified coarsen to always keep the end nodes to prevent `coarsen` from creating gaps in the mesh
- Make `overwritebcs` compatible with an array of data

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
- coarsen a block with an even number of cells. 
- use `overwritebcs` to add an array of data to a bc

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

marcomangano reviewed (approved):
I am not really getting the code changes but if I understand the description correctly this is a nice and long-awaited fix! We don't need to worry about the surface nodes / layers to be necessarily `2n + 1` before coarsening, right?

marcomangano reviewed (commented):
I am not really getting the code changes but if I understand the description correctly this is a nice and long-awaited fix! We don't need to worry about the surface nodes / layers to be necessarily 2n + 1 before coarsening, right?

marcomangano commented:
(I accidentally approved, ofc the conflicts must be fixed!)

sseraj commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -1602,34 +1616,46 @@ def coarsen(self):
+                boco.coarsen(idx_dim)
             for b2b in self.B2Bs:
-                b2b.coarsen(2)
+                b2b.coarsen(idx_dim)
```
Shouldn't the `boco` and `b2b` coarsen calls be inside the above if statement?

joanibal replied:
fixed in the new verison

sseraj commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -1602,34 +1616,46 @@ def coarsen(self):
+        # the number of points in that direction isn't odd
+        for idx_dim in range(3):
+            end_fine_slicer = copy.deepcopy(fine_slicer)
+            end_coarse_slicer = [s] * 3
```
I don't really understand this part. It seems like the end slicers just need to be lists or tuples of dimension 3, not `slice` objects.

joanibal replied:
You could probably do it without slicers, but I think this is cleaner. You would need to add if statements for each dimension.  
```
if idx_dim==0:
    fine_mapping[0] = -1
    new_coords[-1, :, :] = coords[fine_mapping]
elif idx_dim==1:    
    fine_mapping[1] = -1
    new_coords[:,-1, :] = coords[fine_mapping]
    
```

anilyil replied:
This looks fine to me; I agree that the slicer approach is a bit cleaner so sure. 

It also looks like the edge cases of corners would also be handled properly, but I am not 100% sure I could imagine the issues that might come up. As long as you thought about the edge corners getting handled properly (e.g. coarsening on both i, j, and k with an even number of cells on all directions), then this is fine.

sseraj commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -2288,9 +2314,18 @@ def addBocoDataSet(self, bocoDataSet):
         """Coarsen the range of the BC along the specified direction"""
+
         for j in range(2):
             self.ptRange[direction, j] = (self.ptRange[direction, j] - 1) // 2 + 1
```
Does this need to change like in the `b2b` function?

joanibal replied:
yep, fixed in the new version

sseraj commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -2288,9 +2314,18 @@ def addBocoDataSet(self, bocoDataSet):
+                for dir_arr in data_set.dirichletArrays:
+                    dir_arr.dataDimensions[0] = numpy.prod(self.ptRange[:, 1])
+                    data_size = numpy.prod(dir_arr.dataDimensions)
+                    dir_arr.dataArr = dir_arr.dataArr[:data_size]
```
So this will just truncate the data, not take every other element?

joanibal replied:
That is right. The data is stored in a 1-D array, at least for all our meshes. I have added those assumptions as comments.

anilyil replied:
I could not see the comment you mentioned btw. If we go with the truncate option we can either add it in the docstring of the coarsen method and maybe even add a warning if we do it.

Would it be too difficult to also take every other element (or even do some sort of restriction operation) instead of truncating it? I would expect for most use cases of the array BCs, you would set them from the python script anyways, but doing it like this looks like its digging a rabbit hole for an incoming student picking up a mesh with an array BC, coarsening it, and then that changing the entire BC value. I get that it may become difficult figuring out the 1d array indexing but it looks like it would be worth the effort if we do it now rather than someone else later.

joanibal replied:
Alright,  I updated it to take every other element.

joanibal commented:
Yeah, I have some updates that need to be pushed

anilyil commented:
What is the current status of this PR?

joanibal commented:
The PR is ready for review again

joanibal commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -376,13 +375,28 @@ def overwriteBCs(self, bcFile):
+                            for j in range(i, len(aux)):
+                                if aux[j].isnumeric():
+                                    dataArr.append(aux[j])
+                                    i += 1
```
addressed by new changes

anilyil commented on cgnsutilities/cgns_utils.py:
```diff
@@ -253,9 +253,7 @@ def get_parser():
-    dataArrN      - the actual data for the property
-
-Note that only scalar values are supported at the moment for dataArrN.
+    dataArrN      - the actual data for the property. either a scalar or a nodal array.
```
Does it make sense to also say how the nodal data should be ordered here?

anilyil reviewed (commented):
I also reviewed to accelerate this PR a bit. The changes look great in general, just a few comments from me.

anilyil commented:
This pr is fine for me as long as @sseraj is also happy. I can approve after he takes a look again because he had a few comments earlier.

eirikurj commented on cgnsutilities/cgnsutilities.py:
```diff
@@ -1631,34 +1663,46 @@ def coarsen(self):
-        self.dims[1] = self.coords.shape[1]
-        self.dims[2] = self.coords.shape[2]
+
+        # the new dimensions are half rounded up of the old dimensions
+        new_dims = copy.deepcopy(self.dims)
+        for i in range(3):
+            if self.dims[i] > 2:
+                new_dims[i] = (self.dims[i] + 1) // 2
+        new_coords = numpy.zeros((new_dims[0], new_dims[1], new_dims[2], 3))
```
Should we add some type of warning or message that is printed to let the user know if a particular dimension could not be "perfectly" coarsened? The intent seems that when coarsening for instance 1x4 `|-|-|-|` would be coarsened to 1x2 `|-|--|` with a larger second cell. See the attached figure.

joanibal replied:
A message has been added when a dimension cannot be evenly coarsened.

The shortening issue when coarsening from 4 or 3 to 2 has now been fixed too.

eirikurj reviewed (changes requested):
Sorry for the late review on this. Please see my comments. I think we should improve the testing a bit more to include all edge cases, and we should also include/update the examples. There seems to be still some issue. The attached figure shows two cubes (the first column) coarsened (the second column). The second case "looses" the last part and becomes "shorter"
![cube](https://user-images.githubusercontent.com/36180221/144236342-3de25df7-e4c7-453c-ac9e-17da137caf3f.png)

joanibal commented:
Thanks for finding that bug @eirikurj!
I have fixed it and added a test case to make sure it doesn't break in the future. 
It now coarsens the 3x4x2 block to a 2x2x2 block without shortening it.

sseraj commented on cgnsutilities/cgns_utils.py:
```diff
@@ -253,9 +253,7 @@ def get_parser():
-    dataArrN      - the actual data for the property
-
-Note that only scalar values are supported at the moment for dataArrN.
+    dataArrN      - the actual data for the property. either a scalar or a flattened nodal array. if an array is passed, the solver will convert the 1D array to the (possibly) 2D BC face.
```
I think this is the only unaddressed comment

joanibal replied:
done

sseraj reviewed (commented):
Thanks for these changes. This looks good to me as long as @eirikurj's comments are all addressed

joanibal commented:
I fixed the typo

sseraj commented:
Could you also fix the style checks?

joanibal commented:
@nwu63 Do you know why this is being blocked by codecov?

sseraj commented:
> @nwu63 Do you know why this is being blocked by codecov?

Merging is blocked because @eirikurj requested changes. The codecov check is not required

joanibal commented:
Ah, thank you

sseraj commented:
@eirikurj Could you take a look at this again when you get the chance?

eirikurj commented:
This looks good! One minor thing @joanibal did you plan to add this to the examples?

joanibal commented:
I put the new mesh file in the examples folder since that is where the other mesh is. 
Should I move it somewhere else?

eirikurj commented:
@joanibal sorry, the location of the mesh is fine. What I meant is to make an example script that applies the coarsening to that mesh. Same for the overwriting of the BCs (there is one already that could be updated). I know its kind of redundant (and maybe self-explanatory for coarsening), but I think we should try to provide simple examples of all functions and features of the tool, without having to dig through testing code, or relying on extensive examples in doc strings that may end up being inconsistent with what is actually implemented.

eirikurj commented:
@joanibal I had a second look at `coarsen` and ran into some issues. It seems that the BCs have incorrect PointRange.
See the figure below, using mesh `717_wl_L2.cgns` found in the examples directory.
Let's also make sure to add tests that check that these properties are consistent after coarsening.
![cgns_coarsen](https://user-images.githubusercontent.com/36180221/150801160-70f08df8-daed-4431-b904-6aeb44dcf0cf.png)

joanibal commented:
oof, that was a big testing coverage miss on my part. 
I have fixed the bug and added a check to the coarsen test for each boco point range.

I have also added an example of how to use the bcoverwrite with an array of bcdata

sseraj commented:
Can you fix the flake8 checks?

joanibal commented:
I'm not sure why the pylint style check is failing.

sseraj commented:
> I'm not sure why the pylint style check is failing. It doesn't fail when I run it locally.

Are you using the .pylintrc file from our .github repo? In any case, the pylint check is not currently required to merge.

marcomangano commented:
@anilyil can you approve/merge this? I don't feel 100% confident to check all the changes

sseraj commented:
@eirikurj has to approve before this can be merged because of the requested changes
